design-proposals/shadow-node.md

```diff
@@ -0,0 +1,286 @@
+- Prevent virt-handler from modifying non-kubevirt-owned fields on the node. This includes not being able to change the spec or any label/annotation which are not strictly kubevirt owned.
+
+## Non Goals
+
+Redesign/optimize existing virt-hanlder features to not update/patch the node in the first place.
+
+## Definition of Users
+
+Not user-facing change
+
+## User Stories
+
+Not user-facing change
+
+## Repos
+
+Kubevirt/Kubevirt
+
+# Design Options
+
+## Background
+In order to easy review and understand the different options, this section will describe different usage of Node object in the virt-handler.
+
+### Node writers on virt-handler
+Kubevirt sets ~170 labels on each node. This is set by the following entities
+1. virt-handler [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/cmd/virt-handler/virt-handler.go#L185) the node with `kubevirt.io/schedulable=false` at the start and the termination (by SIGTERM only) of the virt-handler process to indicate (best-effort) that the node is not ready to handle VMs.
+
+2. virt-handler/heartbeat:
+  * [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L96) the node with `kubevirt.io/schedulable=false` when the stop channel closes (i.e. virt-handler exits).
+  * [labels and annotates](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/heartbeat/heartbeat.go#L139) the node, updating NodeSchedulable, CPUManager, KSMEnabledLabel labels and VirtHandlerHeartbeat KSMHandlerManagedAnnotation annotations once per minute.
+    * On the other side - Virt-controller's node-ctonroller watches which monitors the heartbeat annotation.
+
+3. Virt-handler/node labeller [labels](https://github.com/kubevirt/kubevirt/blob/5b61a18f4c3b6c549d50fd201664c21720e3dbe2/pkg/virt-handler/node-labeller/node_labeller.go#L254) various information which are mostly static in nature: cpu model, hyperv features, cpu timer, sev/-es, RT...
+  * This is triggered both by interval (3 min) and on changes of cluster config.
+
+## Solution option #1: ShadowNode CRD
+The security issue is being addressed by introducing a new light weight CRD: "ShadowNode" for each Kubernetes node.
+Virt-handler will keep read access to nodes but will lose write access. Virt-handler will instead write to the ShadowNodes object, and this will subsequently be selectively copied to the node by a new controller. This controller will be the only entity who writes to the node object.
+
+Pros:
+- Since the new controller runs virt-controller context which only runs on the control-plane nodes, the risk of taking over the cluster is reduced.
```

Comment:
Are you sure this statement is correct?

On an openshift cluster, I see:
```bash
> k get pod -n openshift-cnv virt-controller-84d87bfdb6-8w55q -o yaml | grep nodename -i
  nodeName: cnvqe-14.lab.eng.tlv2.redhat.com
> k get pod -n openshift-cnv virt-controller-84d87bfdb6-wzkgf -o yaml | grep nodename -i
  nodeName: cnvqe-15.lab.eng.tlv2.redhat.com
> k get nodes
NAME                                      STATUS   ROLES                  AGE     VERSION
cnvqe-14.lab.eng.tlv2.redhat.com          Ready    worker                 7d22h   v1.28.7+f1b5f6c
cnvqe-15.lab.eng.tlv2.redhat.com          Ready    worker                 7d22h   v1.28.7+f1b5f6c
master-0.cnvcl6.lab.eng.tlv2.redhat.com   Ready    control-plane,master   7d23h   v1.28.7+f1b5f6c
master-1.cnvcl6.lab.eng.tlv2.redhat.com   Ready    control-plane,master   7d23h   v1.28.7+f1b5f6c
master-2.cnvcl6.lab.eng.tlv2.redhat.com   Ready    control-plane,master   7d23h   v1.28.7+f1b5f6c
zeus21.lab.eng.tlv2.redhat.com            Ready    worker                 7d22h   v1.28.7+f1b5f6c

```

if indeed virt-controller is running on a worker node, I'm not sure I understand how the security risk is being reduced with this solution.

What am I missing?

Reply:
If you are correct, then it is not. I need to recheck this.
Thanks for this notice!

Reply:
You should not run kubevirt CP on worker nodes. That is a scurity issue on its own.

Reply:
for example: if you take over virt-controller, you can create a pod in every namspace with an image of your choosing, mounting any secret you like and dumping it.

Reply:
> You should not run kubevirt CP on worker nodes.

Interesting. I wasn't aware of it. Probably since I don't think I've even seen kubevirt running this way.

Quick questions:
1. Do you think we should default our infra components to run on master nodes?
2. Is this a common thing to do for operators? Are you aware of a k8s operator that runs pods on master nodes?

Reply:
@iholder101 after checking, it looks like this is configurable on kubevirt CR via node-placement (see [docs](https://kubevirt.io/user-guide/operations/installation/#restricting-kubevirt-components-node-placement)) - however I don't think this configuration comes out of the box (relying only 2 samples of Itamar's cluster seen above and the SNO cluster I'm using..).

@rmohr - if this is a real concern - then 
- shouldn't node-placement configuration for infra components come out of the box? 
- in the end - can we, for the purpose of this proposal, rely on the assumption the virt-controller is indeed only deployed on control-plane nodes?

Reply:
I think it would be a correct general improvement.

But independent how we install by default, if there is a separation between CP nodes and workload then kubevirt must follow that sepration or it is weakening security. For edge deployments they are typically not separated, but even there we could just follow the CP node roles and install fine.

Reply:
Just one thing regarding to openshift, I think I remember vaguely that openshift had an infra node pool in addition to CP nodes where also no workloads are scheduled and that there was a plan to schedule them there (which may be ok too). Does my memory trick me there?

Reply:
not that I know of, [docs](https://docs.openshift.com/container-platform/4.15/nodes/nodes/nodes-nodes-creating-infrastructure-nodes.html) don't seem to mention it either

Reply:
The infra nodes concept is optional in OpenShift.

Reply:
@iholder101 you grep only shows that `nodeName` is provided in the yaml.
But `nodeName` is a field in the spec, thus provided by somebody, not disclosed by the system.
IMU Kube is still not disclosing the actual host name, but only host ip. I'm not sure why, as security by obscurity is no security - but it's still the way it is.

For reference https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodStatus

Reply:
> @iholder101 you grep only shows that `nodeName` is provided in the yaml.
> But `nodeName` is a field in the spec, thus provided by somebody, not disclosed by the system.
> IMU Kube is still not disclosing the actual host name, but only host ip. I'm not sure why, as security by obscurity is no security - but it's still the way it is.

I think there are two things to consider:

 1. I don't think that the node name has a relevancy here, only that our CP runs on non-CP nodes here is relevant. Getting access to worker nodes with compromised pods is much easier than getting access to CP nodes. Opening attack surfaces via angles like mentioned above if our CP runs there.
 2. While, again, node names don't play a role here, I can always get access to all nodes names where I can schedule by simply creating a daemonset and checking the pod status of all created pods. But maybe I don't get the point you try to make.

Reply:
> The infra nodes concept is optional in OpenShift.

If so, I would highly recommend fixing the CP deployment for OpenShift Virtualization.

Reply:
I forgot to update that this PR is ongoing to solve this issue: https://github.com/kubevirt/kubevirt/pull/11659.
It's still facing a few problems (regarding network lanes failures), but I'm actively working on it.

@fabiand 
> But nodeName is a field in the spec, thus provided by somebody, not disclosed by the system.

Unfortunately this is not correct. The `.spec.nodeName` fields acts as both a spec and a status field de-facto, and it will always be populated when the pod lands on a node.